### ISSUE TITLE
Authentication Permissions - Failure

### ISSUE THREAD

**Issue by coda4096** [mongodb-labs/mongo-connector#662]
I am trying to get mongo-connector working with authentication (mongo to ES). Everything was working fine until I added keyfile and SCRAM-SHA-1 authentication to the mongo sharded cluster.

I created a user 'backup' with the role 'backup' as per this page [(https://github.com/mongodb-labs/mongo-connector/wiki/Usage-with-Authentication)]

` { "_id" : "admin.backup", "user" : "backup", "db" : "admin", "credentials" : ... , "roles" : [ { "role" : "backup", "db" : "admin" } ] }`

When I connect to a mongos as the user 'backup' and run the command **db.runCommand("listShards")**, I get a list of shards.

This is my connector parameters:
`mongo-connector -m mdb1:27017 -t es1:9200 -d elastic2_doc_manager --continue-on-error --admin-username backup --password password --logfile /var/log/mongo-connector/mongo-connector.log`

However mongo-connector fails with the error  **OperationFailure: not authorized on admin to execute command { listShards: 1 }**

I even tried giving the 'backup' user the roles 'clusterManager' and 'readAnyDatabase' 

`{ "_id" : "admin.backup", "user" : "backup", "db" : "admin", "credentials" :..., "roles" : [ { "role" : "backup", "db" : "admin" }, { "role" : "readAnyDatabase", "db" : "admin" }, { "role" : "clusterManager", "db" : "admin" } ] }`

but mongo-connector wont even start, giving the error **OperationFailure: Authentication failed.**

Can someone please help. Is this a bug?

**Comment by coda4096** [comment]
FYI, these are the output from my mongos commands

`db.getSiblingDB("admin").auth("backup", "password" )`

```
db.runCommand("listShards")
{
	"shards" : [
		{
			"_id" : "shard001",
			"host" : "shard001/mdb1:27020,mdb2:27020",
			"state" : 1
		},
		{
			"_id" : "shard002",
			"host" : "shard002/mdb3:27020,mdb4:27020",
			"state" : 1
		}
	],
	"ok" : 1
}
```

**Comment by coda4096** [comment]
The full error from mongo-connector.log

2017-02-21 13:11:49,208 [ERROR] mongo_connector.util:96 - Fatal Exception
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/mongo_connector/util.py", line 94, in wrapped
    func(*args, **kwargs)
  File "/usr/lib/python2.7/site-packages/mongo_connector/connector.py", line 370, in run
    'listShards')['shards']:
  File "/usr/lib/python2.7/site-packages/mongo_connector/util.py", line 78, in retry_until_ok
    return func(*args, **kwargs)
  File "/usr/lib64/python2.7/site-packages/pymongo/database.py", line 494, in command
    codec_options, **kwargs)
  File "/usr/lib64/python2.7/site-packages/pymongo/database.py", line 406, in _command
    parse_write_concern_error=parse_write_concern_error)
  File "/usr/lib64/python2.7/site-packages/pymongo/pool.py", line 419, in command
    collation=collation)
  File "/usr/lib64/python2.7/site-packages/pymongo/network.py", line 116, in command
    parse_write_concern_error=parse_write_concern_error)
  File "/usr/lib64/python2.7/site-packages/pymongo/helpers.py", line 210, in _check_command_response
    raise OperationFailure(msg % errmsg, code, response)
OperationFailure: not authorized on admin to execute command { listShards: 1 }

**Comment by coda4096** [comment]
2017-02-21 14:10:12,237 [ALWAYS] mongo_connector.connector:52 - Starting mongo-connector version: 2.5.0
2017-02-21 14:10:12,238 [ALWAYS] mongo_connector.connector:52 - Python version: 2.7.5 (default, Aug  2 2016, 04:20:16) 
[GCC 4.8.5 20150623 (Red Hat 4.8.5-4)]
2017-02-21 14:10:12,238 [ALWAYS] mongo_connector.connector:52 - Platform: Linux-3.10.0-514.el7.x86_64-x86_64-with-redhat-7.3-Maipo
2017-02-21 14:10:12,238 [ALWAYS] mongo_connector.connector:52 - pymongo version: 3.4.0
2017-02-21 14:10:12,256 [ALWAYS] mongo_connector.connector:52 - Source MongoDB version: 3.4.1
2017-02-21 14:10:12,256 [ALWAYS] mongo_connector.connector:52 - Target DocManager: mongo_connector.doc_managers.elastic2_doc_manager version: 0.3.0
2017-02-21 14:10:12,274 [ERROR] mongo_connector.util:96 - Fatal Exception
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/mongo_connector/util.py", line 94, in wrapped
    func(*args, **kwargs)
  File "/usr/lib/python2.7/site-packages/mongo_connector/connector.py", line 398, in run
    hosts, replicaSet=repl_set)
  File "/usr/lib/python2.7/site-packages/mongo_connector/connector.py", line 299, in create_authed_client
    client['admin'].authenticate(self.auth_username, self.auth_key)
  File "/usr/lib64/python2.7/site-packages/pymongo/database.py", line 1048, in authenticate
    connect=True)
  File "/usr/lib64/python2.7/site-packages/pymongo/mongo_client.py", line 505, in _cache_credentials
    sock_info.authenticate(credentials)
  File "/usr/lib64/python2.7/site-packages/pymongo/pool.py", line 523, in authenticate
    auth.authenticate(credentials, self)
  File "/usr/lib64/python2.7/site-packages/pymongo/auth.py", line 470, in authenticate
    auth_func(credentials, sock_info)
  File "/usr/lib64/python2.7/site-packages/pymongo/auth.py", line 450, in _authenticate_default
    return _authenticate_scram_sha1(credentials, sock_info)
  File "/usr/lib64/python2.7/site-packages/pymongo/auth.py", line 201, in _authenticate_scram_sha1
    res = sock_info.command(source, cmd)
  File "/usr/lib64/python2.7/site-packages/pymongo/pool.py", line 419, in command
    collation=collation)
  File "/usr/lib64/python2.7/site-packages/pymongo/network.py", line 116, in command
    parse_write_concern_error=parse_write_concern_error)
  File "/usr/lib64/python2.7/site-packages/pymongo/helpers.py", line 210, in _check_command_response
    raise OperationFailure(msg % errmsg, code, response)
OperationFailure: Authentication failed.

**Comment by ShaneHarvey** [comment]
This is indeed a subtle bug introduced in https://github.com/mongodb-labs/mongo-connector/pull/563. We changed a find on `config.shards` into a call to `listShards` assuming that it would have no change in behavior. Unfortunately (and annoyingly), the `backup` role has privileges to read the list of shards in the `config.shards` collection but, as you can see, does not have the privilege to run the `listShards` command. I'll revert this change to fix the problem in the upcoming 2.5.1 bug-fix release. 

In the meantime, you will need to grant the mongo-connector user the `backup` AND `clusterMonitor` roles.

An important point that is not yet mentioned in the documentation is that the user must be created on a mongos and all the shards. This enables mongo-connector to authenticate to the cluster as a whole and to each shard individually.
